File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


func diameterOfBinaryTree(root *TreeNode) int {
    diameter := 0

    var height func(node *TreeNode) int

    height = func(node *TreeNode) int {
    if node == nil {
        return 0
    }

    left := height(node.Left)
    right := height(node.Right)

    currdiameter := left + right

    if currdiameter > diameter {
        diameter = currdiameter
    } 

    currheight := left

    if right > currheight {
        currheight = right
    }

    return currheight +1;


    }
    height(root);

    return diameter

}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int diameter = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        height(root);
        return this.diameter;
    }

    private int height(TreeNode node) {

        if (node == null) return 0;

        int left = height(node.left);
        int right = height(node.right);

        this.diameter = Math.max(this.diameter, left + right);

        return Math.max(left, right) +1;

    }
}